### PR TITLE
Search functionality for containers

### PR DESCRIPTION
**Pull request by blapid**
As discussed here: https://groups.google.com/forum/#!topic/construct3/7abbfuCGAUI

I added *search* and *search_all* for searching specific tags inside containers. *search* only returns the first occurrence, *search_all* returns all of them.
This should allow users to easily check their Constructs for specific data.
I've implemented some basic tests, they probably dont cover all of the end-cases so if you see anything thats missing, let me know.

As a side note, it could be interesting adding something similar to XPath for more sophisticated searches as done here: http://docs.python.org/2/library/xml.etree.elementtree.html#elementtree-xpath

### CODE REVIEW

File: construct/lib/container.py [hunk 1]
```diff
@@ -70,6 +70,35 @@ def copy(self):
...
+                    else:
+                        return self[key]
+                if type(self[key]) == Container or type(self[key]) == ListContainer:
+                    ret = self[key]._search(name, search_all)
+                    if ret is not None:
+                        if search_all:
+                            items.extend(ret)
+                        else:
+                            return ret
+            except:
```
Comment (arekbulski) [inline construct/lib/container.py]: What is the try/except for and why is it so broad? It should be error specific.

**Comment by arekbulski** [comment]
What is the try/except for and why is it so broad? It should be error specific.

**Comment by arekbulski** [comment]
Would you be willing to add regex into search functionalty? BTW credited you in the docs site.

**Comment by blapid** [comment]
Hi,
Sorry for the delay in answering.
1. About the try/except, I can't really remember , it was a long time ago :) From briefly looking at it I assume it's meant to catch KeyErrors from the `self[key]` calls but I'm not 100% sure.
2. About regex functionality, do you mean simply regex instead of `key == name`? If so yeah I could do that, but I probably wont have time to work on it until November.

**Comment by arekbulski** [comment]
Only 2. No hurry.

**Comment by arekbulski** [comment]
Btw you won't believe how much C2 changed.

**Pull request merged**